src/inmanta/agent/agent_new.py

```diff
@@ -0,0 +1,274 @@
+           see _init_endpoint_names
+        :param agent_map: the agent map for this agent to use
+        :param code_loader: do we enable the code loader (used for testing)
+        :param environment: environment id
+        """
+        super().__init__("agent", timeout=cfg.server_timeout.get(), reconnect_delay=cfg.agent_reconnect_delay.get())
+
+        self.thread_pool = ThreadPoolExecutor(1, thread_name_prefix="mainpool")
+        self._storage = self.check_storage()
+
+        if environment is None:
+            environment = cfg.environment.get()
+            if environment is None:
+                raise Exception("The agent requires an environment to be set.")
+        self.set_environment(environment)
+        assert self.environment is not None
+
+        self.executor_manager: executor.ExecutorManager[executor.Executor] = self.create_executor_manager()
+        self.scheduler = ResourceScheduler(self.environment)
+        self.working = False
+
+    def create_executor_manager(self) -> executor.ExecutorManager[executor.Executor]:
+        # To override in testing
+        return forking_executor.MPManager(
+            self.thread_pool,
+            self.sessionid,
+            self.environment,
+            config.log_dir.get(),
+            self._storage["executor"],
+            LOGGER.level,
+            False,
+        )
+
+    async def stop(self) -> None:
+        await super().stop()
+
+        if self.working:
+            await self.stop_working()
+
+        threadpools_to_join = [self.thread_pool]
+
+        await self.executor_manager.join(threadpools_to_join, const.SHUTDOWN_GRACE_IOLOOP * 0.9)
+
+        self.thread_pool.shutdown(wait=False)
+
+        await join_threadpools(threadpools_to_join)
+
+    async def start_connected(self) -> None:
+        """
+        This method is required because:
+            1) The client transport is required to retrieve the autostart_agent_map from the server.
+            2) _init_endpoint_names() needs to be an async method and async calls are not possible in a constructor.
+        """
+        await self.add_end_point_name(AGENT_SCHEDULER_ID)
+
+    async def start(self) -> None:
+        # cache reference to THIS ioloop for handlers to push requests on it
+        self._io_loop = asyncio.get_running_loop()
+        await super().start()
+
+    async def start_working(self) -> None:
+        """Start working, once we have a session"""
+        # Todo: recycle them when we restart
+        if self.working:
+            return
+        self.working = True
+        await self.executor_manager.start()
+        await self.scheduler.start()
+
+    async def stop_working(self) -> None:
+        """Start working, once we have a session"""
+        if not self.working:
+            return
+        # Todo: recycle them when we restart
+        self.working = False
+        await self.executor_manager.stop()
+        await self.scheduler.stop()
```

Comment:
Not important right now, but I would expect the order to be mirrored compared to the `start_working` method i.e. stopping the scheduler first